### ISSUE TITLE
[Question] Extract BamlRecordReader feasability

### ISSUE THREAD

**Issue by yowl** [dotnet/wpf#10165]
Hi, I contribute to the NativeAOT-LLVM project which is a NAOT compiler for LLVM and WebAssembly.  I am currently bringing up a WPF applicaation in the browser using Avalonia's XPF bridge and have a question.  As you may know NAOT is an AOT compiler that uses roots to establish a dependency tree of required types and methods that need compiling.  WPF presents a challenge in that a some types/properties (get/set methods) may only be used in Xaml or Baml.  So it would be advantageous from the goal of gathering dependencies to read the Baml resources and root the necessary types and methods.  To that end I was looking at BamlRecordReader.cs and wondering if it could be extracted from the PresentationFramework assembly.  Most things are possible, but is this going to be really hard do you think?

Thanks!

**Comment by lindexi** [comment]
@yowl  I firmly believe that this task is feasible. In fact, I think it can be accomplished with just a simple Roslyn analyzer. If you need any assistance from me, I would be more than happy to help.

**Comment by yowl** [comment]
Thanks for the offer of help!  How did you envisage this working given the constraints of a Roslyn analyzer ?

**Comment by lindexi** [comment]
@yowl How about add the Xaml file to Additional file and then analyze it?

**Comment by yowl** [comment]
In many cases we do not have xaml, just the dlls.  The inputs to this process would be dlls and the output would be a trimmer xml file (or modifications to ItemGroups in msbuild, but that sounds more complicated).

**Comment by lindexi** [comment]
@yowl I do not know how to read the embed assembly resource from IAssemblySymbol.

**Comment by maxkatz6** [comment]
[ICSharpCode.BamlDecompiler](https://github.com/icsharpcode/ILSpy/tree/master/ICSharpCode.BamlDecompiler) is an independent parser that could be used to analyze dependencies and emit trimmer metadata before it's read. No idea how feasible this idea is though. Specifically, BAML lacks information about bindings anyway, as WPF doesn't have any concept of compiled bindings like Avalonia/MAUI or xBind like UWP.

**Comment by yowl** [comment]
Yes bindings are still a hole, but if `DataContext` is used, then it could be partially filled, the remaining items, like hierarchical grids would still be holes and would have to be filled manually.  I have something working for the other stuff using IlSpy, testing it now.

**Comment by lindexi** [comment]
Thank you @maxkatz6 

@yowl  I write the demo code about read the baml resource from assembly in IncrementalGenerator. You can find my code in https://github.com/lindexi/lindexi_gd/tree/ba3dba2934e3c56e6efadc9ea91ad3f2f25987b8/Roslyn/RerhabaikaKaloharfo
